### ISSUE TITLE
AppSync, AWS Amplify and SSR

### ISSUE THREAD

**Issue by romainquellec** [aws-amplify/amplify-js#1613]
AppSync, AWS Amplify and SSR : 
Hi ! Is Amplify (GraphQL Client) is SSR compatible ? Can't find any information on docs right now. Thanks !

**Comment by undefobj** [comment]
We have a tutorial published for this with Next.js: https://medium.com/open-graphql/ssr-graphql-apps-with-next-js-aws-appsync-eaf7fbeb1bde

**Comment by romainquellec** [comment]
Well, in this example, you re not using amplify.

**Comment by romainquellec** [comment]
The answer is : no, you can't.

**Comment by Enalmada** [comment]
@romainquellec Can you elaborate?  I am considering building an AppSync react native app for a client and an admin interface to manage things is a requirement.  A Next.js Amplify AppSync admin client "backend" would be ideal.  Is Amplify with Next.js really not possible?

**Comment by romainquellec** [comment]
@undefobj ?

**Comment by littleone2018** [comment]
i only want seo, can amplify work with sth like prerender.io?

**Comment by ajhool** [comment]
Any luck with amplify and next.js? @Enalmada @romainquellec

**Comment by romainquellec** [comment]
I cant help, I'm now using apollo and appsync for the GraphQL part.

**Comment by skdigital** [comment]
Its a real shame that nextjs is not compatible with AWS Amplify, its a massive short fall in my eyes.

**Comment by kjetilge** [comment]
I'm having the same issue. Maybe it would it be possible/easy to trigger prerendering of routes from a lambda function ? Perhaps using PrerenderSPAPlugin ?

**Comment by romainquellec** [comment]
The specific GraphQL part was not SSR compatible (months ago). 
I'm using Auth and Storage with Next with no problem (for now).

```
    "@aws-amplify/auth": "^1.2.15",
    "@aws-amplify/core": "^1.0.22",
    "@aws-amplify/storage": "^1.0.25",

...

    "apollo-cache-inmemory": "^1.4.2",
    "apollo-client": "^2.4.12",
    "apollo-link-http": "^1.5.11",
    "aws-appsync": "^1.7.1",
```

is OK.

**Comment by romainquellec** [comment]
And you still could load some components with <NoSSR> component with React if something is broken.

**Comment by ngocketit** [comment]
I'm using Amplify with Appsync and Nextjs and it works pretty ok. I had a setup where I use Api key on server and Cognito user pool on client. SSR is for public, anonymous user so you don't need to handle authentication etc. For all the protected pages that require user login, I render it on client.

**Comment by romainquellec** [comment]
Closing this. There is (at least) one issue on SSR improvement for aws-amplify.

**Comment by ajhool** [comment]
I'm not sure why this issue was closed. This is an ongoing issue that I think a lot of people will stumble upon as nextjs and amplify grow in popularity.

**Comment by skdigital** [comment]
I agree this shouldn't be closed.. I, along with many others are holding our breath for amplify support for server rendered apps.

**Comment by ajhool** [comment]
@skdigital what are you looking for in particular? I have a nextjs workaround that I think is robust, although the workaround is just to deter Amplify from doing anything on the server

**Comment by skdigital** [comment]
@ajhool 

I am looking to use nextjs and host via zeit now.

Use only Amplify for Auth, Storage, and Appsync realtime DB graphql with offline functionality.

Is this possible and advisable for production? (hosting somewhere else, so we can use nextjs)?

**Comment by ajhool** [comment]
I would highly recommend building a test app, first, that attempts the integrations you plan on using. If you'd like, I can create an open source example to show my integration strategies. I'm using react.

The only problem I've had, so far, is Amplify Auth + next. I have used Storage and Api(Appsync - not realtime) without any issues. All of my Api calls are coming from the client, so I'm not sure if Api/Appsync will work on the server. Especially if the calls require authentication.

One of the main frustrations has been that the Amplify team does not currently fully support not-breaking-SSR** in the Amplify framework. However, the team is also extraordinarily helpful and responsive, they just have not been responsive on the SSR issue; I expect that to change as more people request SSR compatibility. Also, because amplify is an open source project provided by AWS, it seems that the official AWS support center does not actually provide support -- this policy, to me, does not make sense, but the lack of official support is worth considering for a production environment.

The other frustration is that next/now/zeit team is not very responsive or supportive. The software is convenient and worth it, but when it breaks it can be hard to find a fix. Thus, I highly recommend a barebones test app before integrating your entire app (as I tried to do).

**By not-breaking-ssr, I mean that the framework does not need to actual render on the server-side, but it should not break server-side rendering. Currently, simply importing certain Amplify components (eg. `aws-amplify-react/Authenticator`) has broken SSR without any clear error or debug opportunities.

**Comment by gcid12** [comment]
@romainquellec , I also think you should keep this ticket open. There's a lot you can do with Amplify But well supported, well documented, production ready Server Side is a must for a lot of us.  Would love to hear from @undefobj if this is in the roadmap, he's usually very responsive.

**Comment by undefobj** [comment]
SSR support is on our roadmap, and we'd like to do it this year but for transparency it's not in the next few months. It will require a fair amount of work on the internals and potentially some breaking changes so we would like to do it as part of a larger release this summer/fall with other changes. That being said if anyone in the community wishes to fork Amplify and try to get it working, then give feedback on what was required to get things working (even if it's a PR we don't merge) then that would help us accelerate the work.

I've reopened this issue for tracking.

**Comment by ajhool** [comment]
Next.js has added typescript and javascript Amplify examples: https://github.com/zeit/next.js/tree/canary/examples/with-aws-amplify-typescript 

I haven't tested it yet to see what is/isn't working and I'm a little skeptical, but hopefully it's helpful

**Comment by AshleyTharpTL** [comment]
> i only want seo, can amplify work with sth like prerender.io?

I also want to use prerender.io on my site hosted with Amplify but I cannot find a way to do this.

**Comment by CaptainChemist** [comment]
@AshleyTharpTL you might want to try https://www.prerender.cloud/ because I know that works for static sites created with create-react-app so it would likely work here. The idea is you deploy a cloudformation template that sets up a lambda function which will be in front of your website. As requests come in, the lambda will pre-render everything which will give you server-side rendering.

**Comment by gtinkham** [comment]
> I'm using Amplify with Appsync and Nextjs and it works pretty ok. I had a setup where I use Api key on server and Cognito user pool on client. SSR is for public, anonymous user so you don't need to handle authentication etc. For all the protected pages that require user login, I render it on client.

@ngocketit Do you have an example as to how you got this working? 
Does this method still work in the current 1.2.4 version of aws-amplify?
I am currently trying to implement something similar, but I cant seem to get any queries to run without totally breaking my next project.

**Comment by ajhool** [comment]
@gtinkham 

Can't get any queries to run *on your server?*

Have you worked with graphql before?

**Comment by gtinkham** [comment]
@ajhool 
So for clarification, I can get queries to run via apollo client, however, I want to use Amplify as opposed to apollo for my project.
But I have run into several issues including #1613 while trying to integrate aws-amplify 1.2.4 into my next project

**Comment by ajhool** [comment]
@gtinkham 

I recently switched from the amplify client to the apollo client and it has made life so much easier. The apollo client has a few bugs but it beats the amplify client in nearly every way IMO and I've recommended to the amplify team that they should remove the amplify client because things like SSR are not workable and there are existing solutions. I don't actually know if you can use the amplify client on a server -- I do know that amplify officially does not support server-side code.

I've also was finding that simply importing amplify on the server was breaking it -- I've never had a node package break server-side rendering or code as badly as amplify does. The only way I was able to use Amplify with next and SSR was to defer import to the client side using next's dynamic import.

**Comment by mlabieniec** [comment]
@gtinkham what are the issues you are having running amplify with ssr? Are you using next.js? I have run through this tutorial using amplify and API (GraphQL) without any issues using next.js:
https://dev.to/rakannimer/server-side-rendered-real-time-web-app-with-next-js-aws-amplify-graphql-2j49

**Comment by gtinkham** [comment]
@ajhool I have found the same frustrations attempting to bring Amplify into my new project, and normally I would have gone with apollo since that was a breeze to get running, but my requirements dictate that I use Amplify.

I have gotten it to run by using some resolutions to older packages inside amplify that I found on a separate issue: #4305 
I am not sure if there are any large security risks in doing so, but it seems to be running fine now.

@mlabieniec I did not have success with any online tutorial including the current 1.2.4 version of Amplify, but as stated above I think I have a temporary solution until the amplify team is ready to support SSR out of the box.

**Comment by mlabieniec** [comment]
@gtinkham For the navigator you have to add a polyfill line in the next.config.js, as well as a few additional things to render the UI, see here: https://github.com/aws-amplify/amplify-js/issues/3854#issuecomment-554200567

We are working on some doc improvements/getting started for SSR as well as improved support.

**Comment by gtinkham** [comment]
@mlabieniec I added the line for global.navigator in my next.config.js which seems to have fixed my running errors. However, I get that same navigator issue when I try to build my project.
Is there some other place I need to reference navigator to solve that issue?

**Comment by mlabieniec** [comment]
@gtinkham there is a fix for that in the latest amplify which you can get with `npm i --save aws-amplify@next` (2.0). This will fix that issue. However, there is still a build issue with the css in the node_modules (@aws-amplify/ui/dist/style.css) being loaded, which is an issue with next.js in discussion here: https://github.com/zeit/next-plugins/issues/266

Seeing if there is anything we can do to figure out a work-around for this in the meantime, next also has an RFC for this: https://github.com/zeit/next.js/issues/8626

**Comment by mlabieniec** [comment]
@gtinkham You can fix the build by using the following `next.config.js` (this works for both dev and build:

```js
const withCSS = require('@zeit/next-css')
const resolve = require('resolve')
global.navigator = () => null
module.exports = withCSS({
    webpack (config, options) {
        const { dir, isServer } = options
        config.externals = []
        if (isServer) {
          config.externals.push((context, request, callback) => {
            resolve(request, { basedir: dir, preserveSymlinks: true }, (err, res) => {
              if (err) {
                return callback()
              }
              if (
                res.match(/node_modules[/\\].*\.css/)
                && !res.match(/node_modules[/\\]webpack/)
                && !res.match(/node_modules[/\\]@aws-amplify/)
              ) {
                return callback(null, `css ${request}`)
              }
    
              callback()
            })
          })
        }
    
        return config
      }
});
```

**Comment by shawnmmatthews** [comment]
@undefobj are there any updates on this with regards to official support?

**Comment by samstr** [comment]
We want SSR! ✊

**Comment by ragate-tatsuyoshi** [comment]
Basically, Web application should be optimized SEO.
2 months ago, I don't used amplify when I created my product because amplify not support server side render.

**Comment by danielblignaut** [comment]
Would be good to get opinions on this... May seem slightly hacky but I've managed to at least get the Auth category storage running on SSR... From what I can tell, a great step forward for SSR in this library would be allowing us to universally configure the data storage across all categories. for example, in @aws-amplify/core package, storage is configured to use either window.localstorage and if no window.localstorage is available (i.e. on the server) it defaults to just using a js object. Would be great to be able to override this to use cookie storage instead? as we could then use universal cookies on the browser and server in nextJS. Currently, I'm doing something like this (very rough and extremely hacky until https://github.com/aws-amplify/amplify-js/blob/master/packages/core/src/StorageHelper/index.ts is allowed to pass storage object in:

Basically I'm overwriting window.localstorage to a custom storage object, configuring amplify to use that storage object then removing it such that it doesn't mess other dependency configs, etc. This is done on both the server and browser whilst bootstrapping Amplify.

I also need to add config.storage to my config, as the packages/amazon-cognito-identity-js which seems to do a lot of the underlying heavy lifting for the auth category takes a storage parameter (see https://github.com/aws-amplify/amplify-js/blob/master/packages/amazon-cognito-identity-js/src/CognitoUserPool.js )

My storage class then uses a universal client and server-side cookie library called nookies in nextJS.

configureAmplify.js
```js
import config from '@/amplify/aws-exports'
import Amplify from '@aws-amplify/core'
import Auth from '@aws-amplify/auth'
import API from '@aws-amplify/api'
import Storage from '@aws-amplify/storage'
import { UniversalCookieStorage } from '@/lib/cookieStorage'
import { NextPageContext } from 'next'

export default (ctx?: NextPageContext)=> {
	if(ctx) UniversalCookieStorage.setContext(ctx)
	let isServer = false
	let localStorage = null
	//@ts-ignore
	config.storage = UniversalCookieStorage
	if(typeof window === 'undefined') {
		//@ts-ignore
		global.window = {
			localStorage: UniversalCookieStorage
		}
		isServer = true
	}
	else {
		localStorage = window.localStorage
		Object.defineProperty(window, 'localStorage', {
			value: UniversalCookieStorage,
			configurable: true,
			enumerable: true,
			writable: false
		})
	}
	Amplify.configure(config)
	//Amplify.Logger.LOG_LEVEL = 'ERROR'
	Auth.configure(config)
	Storage.configure(config)
	API.configure(config)

	if(isServer) {
		//@ts-ignore
		global.window = undefined
	}
	else {

		Object.defineProperty(window, 'localStorage', {
			value: localStorage,
			configurable: true,
			enumerable: true,
			writable: false
		})
	}
}
```

cookieStorage.ts

```js
import { NextPageContext } from 'next'
import { setCookie, parseCookies, destroyCookie } from 'nookies'


export class UniversalCookieStorage {
	public static ctx: NextPageContext | null = null
	public static maxAge = 30 * 24 * 60 * 60

	public static setContext(ctx: NextPageContext) {
		UniversalCookieStorage.ctx = ctx
	}
	

	/**
	 * This is used to set a specific item in storage
	 * @param {string} key - the key for the item
	 * @param {object} value - the value
	 * @returns {string} value that was set
	 */
	public static setItem(key: string, value: any) {
		if(typeof value != 'string' && typeof value == 'boolean' && typeof value == 'number') {
			value = JSON.stringify(value)
		}
		setCookie(UniversalCookieStorage.ctx, key, value, {
			maxAge: UniversalCookieStorage.maxAge,
			path: '/',
		})
	}

	/**
	 * This is used to get a specific key from storage
	 * @param {string} key - the key for the item
	 * This is used to clear the storage
	 * @returns {string} the data item
	 */
	public static getItem(key: string) {
		const cookies = parseCookies(UniversalCookieStorage.ctx)
		// if(cookies[key] == null) throw new Error(`${key} does not exist in cookies`)

		return cookies[key]
	}

	/**
	 * This is used to remove an item from storage
	 * @param {string} key - the key being set
	 * @returns {string} value - value that was deleted
	 */
	public static removeItem(key: string) {
		const cookies = parseCookies(UniversalCookieStorage.ctx)

		// if(cookies[key] == null) throw new Error(`${key} does not exist in cookies`)
		destroyCookie(UniversalCookieStorage.ctx, key, {
			path: '/'
		})
		return cookies[key]
	}

	/**
	 * This is used to clear the storage
	 * @returns {string} nothing
	 */
	public static clear() {
		const cookies = parseCookies(UniversalCookieStorage.ctx)
		const keys = Object.keys(cookies)

		keys.forEach((key)=> {
			destroyCookie(UniversalCookieStorage.ctx, key)
		})

		return {}
	}
}


```

example component with SSR rendering for a user (i.e. show authenticated page otherwise redirect user)

```js

import React, { useState, useEffect } from 'react'
import dynamic from 'next/dynamic'
import isBrowser from '@/lib/isBrowser'
import withAuthSync from '@/components/auth/withAuthSync'
import { isUserLoggedIn } from '@/lib/user'
import Auth from '@aws-amplify/auth'
import redirect from '@/lib/redirect'
import configureAmplify from '@/lib/configureAmplify'


// const Home: React.FC<{}> = ()=> {
// 	return <div>
// 		<a onClick={async (e: React.MouseEvent<HTMLAnchorElement>)=> {await Auth.signOut(); redirect(process.env.sign_in_url)}}>log out</a>
// 	</div>
// }


// export default withAuthSync(Home, ['dietitian', 'user'])

interface Props {
	isLoggedIn: boolean
}

const Example: React.FC<Props> = (props)=> {

	if(props.isLoggedIn) {
		return <div>Secret content</div>
	}
	else {
		return <div>need to log in!</div>
	}
}

//@ts-ignore
Example.getInitialProps = async (ctx: NextPageContext)=> {
	configureAmplify(ctx)
	let user = null
	let isLoggedIn = false
	try {
		user = await Auth.currentAuthenticatedUser()

		if(user != null) isLoggedIn = true
	}
	catch(e) {
		// console.error(e)
	}

	return {
		isLoggedIn
	}
	
}

export default Example

```

and lastly, _app.js

```js
import React from 'react'
import App, { AppInitialProps, AppProps, AppContext } from 'next/app'
import configureAmplify from '@/lib/configureAmplify'

interface MyAppProps extends AppProps, InitialPageLoadAppProps { 
}

interface InitialPageLoadAppProps extends AppInitialProps {

}

interface MyAppState {
}



class MyApp extends App<MyAppProps, {}, MyAppState> {
	// Only uncomment this method if you have blocking data requirements for
	// every single page in your application. This disables the ability to
	// perform automatic static optimization, causing every page in your app to
	// be server-side rendered.
	
	// public static async getInitialProps(appContext: AppContext) {
	// 	//server-side auth checks and redirects here


	// 	// calls page's `getInitialProps` and fills `appProps.pageProps`
	// 	const appProps = await App.getInitialProps(appContext)

	// 	return { ...appProps }
	// }

	public constructor(props: MyAppProps) {
		super(props)
		this.state = {
		}
		
		configureAmplify()
	}

	public async componentDidMount() {
		try {
			this.setState({
				...this.state,
			})
		}
		catch(e) {
			console.error(e)
		}
	}

	public render() {
		const { Component, pageProps } = this.props
		return <Component {...pageProps} />
	}
}




export default MyApp
```

I haven't played around with the API library and the above as yet but if we can get it to play nice, at least that will let one use SSR for the auth and API categories.... which is my mind will be most of the use cases for SSR anyway. 

Interested to hear feedback.

**Comment by lcx1204** [comment]
@undefobj, do you think SSR could become available next few months? thank you

**Comment by undefobj** [comment]
We don't have any specific timelines but we are actively looking into this.

**Comment by MontoyaAndres** [comment]
Hi, I'm looking for adding Next.js to my apps with Amplify. I don't know if this example works only with create-react-app: https://github.com/aws-samples/aws-amplify-auth-starters/tree/react because it does not use localstorage manually, it just uses the methods that Amplify auth offers, I don't know if these methods are the enough smart to detect if the app is using ssr or not, even with ssg. Algo what @danielblignaut is showing in his example would work great, it's very similar to this: https://spectrum.chat/next-js/general/amplify-auth-and-next-js~b7ef99fe-d6bf-459d-8dd3-d5a04fcbaf89 I don't know if you used it as a reference.

Also, Next.js 9.3 has new methods that are more powerful than `getInitialProps` which are:

`getStaticProps` fetch data over ssg
`getServerSideProps` fetch data over ssr instead of using `getInitialProps`
`getStaticPaths` to control the path and params of ssg

Maybe these methods can help to bring this functionality as soon as possible :)

**Comment by samstr** [comment]
Just deploy to [now](https://zeit.co/). It works perfectly with serverless functions & SSR as it's built by zeit (nextjs). It uses AWS Lambda behind the scenes.

**Comment by judygab** [comment]
> Its a real shame that nextjs is not compatible with AWS Amplify, its a massive short fall in my eyes.

why it's not compatible?

**Comment by judygab** [comment]
> @gtinkham
> 
> I recently switched from the amplify client to the apollo client and it has made life so much easier. The apollo client has a few bugs but it beats the amplify client in nearly every way IMO and I've recommended to the amplify team that they should remove the amplify client because things like SSR are not workable and there are existing solutions. I don't actually know if you can use the amplify client on a server -- I do know that amplify officially does not support server-side code.
> 
> We also were finding that simply importing amplify on the server was breaking the code due to a `window` reference -- I've never had a node package break server-side rendering or code as badly as amplify does (a `window` reference at import time!!!). The only way I was able to use Amplify with next and SSR was to defer import to the client side using next's dynamic import.
> 
> Before the switch, I was using apollo to make public calls from the server and then was using the amplify client on the client.

can you provide and example of app or index file?

**Comment by ericclemmons** [comment]
For reference, this is in `@preview`:
> https://github.com/aws-amplify/amplify-js/issues/5435#issuecomment-655660103

**Comment by ericclemmons** [comment]
This has been addressed as part of #6146!

Please see my synopsis on changes here: https://github.com/aws-amplify/amplify-js/issues/5435#issuecomment-692990722
